### PR TITLE
[FLINK-22676][coordination] The partition tracker stops tracking internal partitions when TM disconnects

### PR DESCRIPTION
**Pull request by jinxing64**
## What is the purpose of the change

When TM disconnects(`JobMaster#disconnectTaskManager`), `JobMasterPartitionTrackerImpl` stops tracking all the partitions the TM ever produced, i.e. the lifecycle of shuffle data is bound to computing resource (TM). It works fine for internal shuffle service, but doesn't for remote shuffle service. Note that if shuffle data is accommodated on remote, the lifecycle of a completed partition is capable to be decoupled with TM, i.e. TM is totally fine to be released when no computing task on it and further shuffle reading requests could be directed to remote shuffle cluster.  
This PR proposes to fix `JobMasterPartitionTrackerImpl` and handle above scenario properly.

## Brief change log

  - Fix `JobMasterPartitionTrackerImpl#stopTrackingPartitionsFor` -- only internal partitions, which are accommodated on TM local, are stopped tracking, 
  - Fix `JobMaster#jobStatusChanged` -- current logic traverses `registeredTaskManagers` and release or promote related partitions, which is incorrect. With above change, external partitions, which are accommodated on external shuffle service, will NOT be released when TM disconnects. Thus release partitions by traversing existing `registeredTaskManagers` might cause leak of external partitions. 
  - Exclude internal partitions when invoke `internalReleasePartitionsOnShuffleMaster`

## Verifying this change

Added new tests in JobMasterPartitionTrackerImplTest

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by guoweiM** [comment]
Thanks @jinxing64 for opening this pr. In general I could understand what we want to fix. But I have a little concern.

In theory, if internal shuffle is not used, `JobMasterPartitionTracker` continues to maintain the mapping of TM to `ResultPartition` may be counter-intuitive. This can also lead to some inconsistencies, such as:

1. In the case of Remote Shuffle, after the job reaches the Terminated state, the `ResultPartition` and the TM to `ResultPartition` mapping will be deleted at the same time, but if the TM is lost, only the TM to `ResultPartition` mapping will be deleted. In this case, it may be meaningless to maintain this mapping, and if this inconsistency continues, it may make it difficult for others to understand the reason for this approach. 

2. Another case is that in the case of Remote Shuffle, after the one task ends, because we continue to maintain the mapping of TM to `ResultPartition`, we might think that `ResultPartition still occupies TM resources at sometime.

I have not yet thought of a specific and complete solution, but a possible direction may be: If JobMasterPartitionTracker has found that ResultPartitionDeploymentDescriptor is not an internal shuffle, don’t maintain the mapping from TM to `ResultPartition` at the beginning.

WDYT?

**Comment by jinxing64** [comment]
Thanks for comments @guoweiM . 
Currently there are two data structures in JobMasterPartitionTrackerImpl:
1. `partitionTable`, which maintains the mapping from tmId to result partitions it produced;
2. `partitionInfos`, which records all the partitions under tracking;

When a tm is gone, it's relative partitions in `partitionTable` will be cleared, which means no partitions is accommodated on it;
At the same time, only tm internal partitions is cleared from `partitionsInfos` but external partitions are kept -- external partitions are available and could be decoupled from the lifecycle from tm;
I think that's where the `inconsistencies` you mentioned came from. And I agree that it take words to explain the underlying mechanism and might be hard for understanding for new commers;

> If JobMasterPartitionTracker has found that ResultPartitionDeploymentDescriptor is not an internal shuffle, don’t maintain the mapping from TM to ResultPartition at the beginning.

I agree with you for the proposal -- `partitionTable` only maintains tm internal partitions from the start and `partitionInfos` records all external and internal partitions. With this change the semantics of below interfaces will be:
1. `boolean isTrackingPartitionsFor(K key);`: if there's internal partitions tracked for the `key`;
2. `boolean isPartitionTracked(ResultPartitionID resultPartitionID)`: if the resultPartitionID is tracked anywhere, no matter internally or externally

**Comment by jinxing64** [comment]
Thanks a lot for the concern @guoweiM ; How do you think? @tillrohrmann and @zhuzhurk 
It will be great to hear comments from you Flink master. I will keep refining this PR if we can reach consensus for the idea.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/TestingJobMasterPartitionTracker.java
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/TestingJobMasterPartitionTracker.java:116]: maybe also add a `Function` for this method to make it customizable?

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterPartitionReleaseTest.java
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterPartitionReleaseTest.java:171]: This verification looks a bit weird. I guess what it is currently checking that the future is completed normally.
Maybe we can manually add some partitions to the tracker before terminating the job, and check whether they are released on job termination.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:92]: Looks to me this invocation is not needed.
Since we now have `partitionTable#getTrackedPartitions()`, maybe we can also fix `AbstractPartitionTracker#stopTrackingPartitionsFor(...)` which calls `partitionTable.stopTrackingPartitions(...)` twice.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:86]: I prefer the name to be `resultPartitionIds` just like other places in this class.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:231]: I think we should not exclude partitions which occupy local resources. Because conceptually external partition can also occupy local resources.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (zhuzhurk) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:113]: Maybe we can add a method `stopTrackingAndHandlePartitions(...)` and reuse it in both`stopTrackingAndReleasePartitions(...)` and `stopTrackingAndReleaseOrPromotePartitions(...)`?
```
    public void stopTrackingAndHandlePartitions(
            Collection<ResultPartitionID> resultPartitionIds,
            BiConsumer<ResourceID, Collection<ResultPartitionDeploymentDescriptor>>
                    partitionHandler) {
        Preconditions.checkNotNull(resultPartitionIds);

        // stop tracking partitions to handle and group them by task executor ID
        Map<ResourceID, List<ResultPartitionDeploymentDescriptor>> partitionsToReleaseByResourceId =
                stopTrackingPartitions(resultPartitionIds).stream()
                        .collect(
                                Collectors.groupingBy(
                                        PartitionTrackerEntry::getKey,
                                        Collectors.mapping(
                                                PartitionTrackerEntry::getMetaInfo, toList())));

        partitionsToReleaseByResourceId.forEach(partitionHandler);
    }
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java:116]: Looks to me that we do not need this block at all because it is testing the removed `stopTrackingAndReleasePartitionsFor()`

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java
Comment (zhuzhurk) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java:266]: I think `contains` is better which requires exact matching of collections.

**Review by zhuzhurk — COMMENTED** [review]
Thanks for opening this PR! @jinxing64 
I think it is in the good direction. There are a few comments and please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/PartitionTable.java
Comment (jinxing64) [reply flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/PartitionTable.java:68]: Sure

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/TestingJobMasterPartitionTracker.java
Comment (jinxing64) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/TestingJobMasterPartitionTracker.java:116]: ok

File: flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterPartitionReleaseTest.java
Comment (jinxing64) [reply flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterPartitionReleaseTest.java:171]: That's much better.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (jinxing64) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:92]: Sure I will remove this invocation ~

But I'm hesitate to think there's issue in `AbstractPartitionTracker#stopTrackingPartitionsFor` -- the idea is to stop tracking by tmID and get corresponding partitionID, then stop tracking by partitionID. Is there redundancy ? Did I get it wrong ?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (jinxing64) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:86]: Sure

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (jinxing64) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:92]: That's better ~

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (jinxing64) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:231]: ok, though I don't find a scenario in which an completed external partition occupies tm resource. Note that if we keep excludePersistentPartitions, shufflemaster might be fed with unknown partitionid

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java
Comment (jinxing64) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java:116]: Yes, I also had such concern but was hesitate to remote the existing testing block -- at least current code tests the process of un-tracking partitions one by one.

Besides `testReleaseCallsWithoutLocalResources` also has the similar issue.
Do you insists?  I can remote the block if so.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java
Comment (jinxing64) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java]: `contains` is sensitive about order while `hasItems` doesn't

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (jinxing64) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java:111]: cool ~~

**Comment by jinxing64** [comment]
wip

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java
Comment (jinxing64) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java:79]: Please review and check this change  -- since `JobMasterPartitionTrackerImpl#partitionTable` only tracks internal partitions, I updated field of `hasLocalResources` as true.

**Comment by jinxing64** [comment]
@zhuzhurk 
Thanks for deep review ~ I updated by your comments ~ please take a look when CI gives green ~~

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java
Comment (zhuzhurk) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImplTest.java]: I think we can remove the duplicate blocks in `testReleaseCallsWithLocalResources()` and `testReleaseCallsWithoutLocalResources()`.
If to test process of un-tracking partitions one by one, I think we should extract the block logic into one method.

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java]: All partitions should have been registered to the `ShuffleMaster` no matter whether it occupies TM resources.
So I think `ShuffleMaster` would not see unknown partition ids.

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/TestingJobMasterPartitionTracker.java
Comment (jinxing64) [reply flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/TestingJobMasterPartitionTracker.java]: resolved

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java
Comment (jinxing64) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/JobMasterPartitionTrackerImpl.java]: that's true

**Comment by jinxing64** [comment]
@zhuzhurk 
I resolved your comments by a following commit. Please take a look when you have time. I will squash the commits once the whole pr is approved.

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing all the comments! @jinxing64 
The change looks good to me.
You can squash the changes now and I will merge it once CI gives green.

**Comment by jinxing64** [comment]
@zhuzhurk 
Thanks for quick response ~ I've squashed the commits.

**Comment by zhuzhurk** [comment]
Rebased onto latest master because the PR base was a bit too old.

**Comment by jinxing64** [comment]
Thanks @zhuzhurk for rebasing ~ please ping me to rebase for convenience next time :)

**Comment by jinxing64** [comment]
Thanks for merging ~~

**Pull request closed without merging**
